### Build models that predict one of two categories provided by X and y matrices. 

In [1]:
%matplotlib inline

In [2]:
%run __init__.py

#### Truncated SVD now available in namespace. 

Use get_dummies, drop first pandas functionality to encode the categories as 0 and 1. 

In [3]:
wiki_df['target'] = pd.get_dummies(wiki_df['Category'], drop_first=True)

In [4]:
wiki_df['Category'].iloc[0], wiki_df['target'].iloc[0]

('Machine Learning', 1)

Machine Learning is encoded as category 1, while Business Software is encoded as category 0. 

In [5]:
y = np.array(wiki_df['target'])

In [6]:
X = wiki_svd_matrix

In [7]:
from sklearn.model_selection import train_test_split

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2, random_state=42)

In [9]:
from sklearn.linear_model import LogisticRegression

In [10]:
log_model = LogisticRegression()
lr = log_model.fit(X_train, y_train)
lr.score(X_train, y_train)

0.97650000000000003

In [11]:
lr.score(X_test, y_test)

0.9760479041916168

Using TFDFI and SVD, predicting a binary class on a corpus of text proves to be a simple matter using only Logistic Regression. The model scores a .976 R2 on test data. 

The goal of this exercise is to be able to predict a class based on feeding the model a sample of text data from wikipedia. 

Below I will test my model to see whether it is able to predict class correctly for a single sample of text. In order to do so, I will transform the sample text using tfdfi and svd, and then use the predict function to predict the category.

In [12]:
sample_text = ml_df.sample(1)

In [13]:
sample_text

Category                       _id  \
252  Machine Learning  5a0d0b218423e1001feda37e   

                                             categoies  page_id  \
252  [Category:All articles lacking in-text citatio...  2203756   

                                                  text          title  
252  {{More footnotes|date=March 2011}}\n{{multiple...  Jaccard index

The chosen sample text here is in the Machine Learning category, cooresponding to numerical category 1. Therefore, the model should predict 1 %97.6 of the time if it is working correctly. 

In [14]:
sample_text['text'] = sample_text['text'].map(str)
sample_text['text'] = sample_text['text'].apply(cleaner)

In [15]:
sample_text.set_index('page_id', inplace=True)

In [16]:
sample_matrix_spare = tfidf_vector.transform(sample_text['text'])
sample_df_tfd = pd.DataFrame(sample_matrix_spare.toarray(),
                                index=sample_text.index,
                                columns=tfidf_vector.get_feature_names())

In [17]:
sample_svd_matrix = SVD.transform(sample_df_tfd)

In [18]:
lr.predict(sample_svd_matrix)

array([1], dtype=uint8)

The prediction is accurate, predicting 1 or Machine Learning as the category of the text. For brevity I will not prove that the model is accurate 97.6% percent of the time. 

Note that if the functionality was desired, the text would need to be loaded, transformed, and then fed to the model each time. For the purposes of this project I will not do so, but could add that functionality later on. 

To check that no data leakage is occuring (some of the wiki text reference the category directly in the text body of the page) I will manually copy and paste a text snippet from a random article, ensuring that no data leakage is happening. If them model truly works as well as it scores, it should accurately predict. 

In [39]:
wiki_df['text'].sample(1)[0]

'multiple issuesrefimprovedateoctober selfpublisheddateoctober advertdateapril infobox software name                    plesk logo                    screenshot              plesk screenshotjpg caption                 something smarter than just plesk screenshot developer               plesk latest release version    latest release date     start date and age operating system        windows linux genre                   web hosting control panel license                 trialware website                 urlpleskcomplesk is a commercial web hosting platform with a control panel that allows a server administrator to set up new websites reseller accounts email accounts and dns entries through a webbased interfaceoriginally designed in novosibirsk russia the hosting automation software was released by plesk inc  and went live first in  overview plesks control panel softwarecontrol panel allows a server system administratoradministrator to set up new websites reseller accounts email accounts

In [32]:
test_text = "dynamic topic models are generative models that can be used to analyze the evolution of unobserved topics of a collection of documents over time this family of models was proposed by david blei and john lafferty and is an extension to latent dirichlet allocation lda that can handle sequential documentsin lda both the order the words appear in a document and the order the documents appear in the corpus are oblivious to the model whereas words are still assumed to be de finettis theoremexchangeable in a dynamic topic model the order of the documents plays a fundamental role more precisely the documents are grouped by time slice eg years and it is assumed that the documents of each group come from a set of topics that evolved from the set of the previous slicetopicssimilarly to latent dirichlet allocationlda and plsa in a dynamic topic model each document is viewed as a mixture of unobserved topics furthermore each topic defines a multinomial distribution over a set of terms thus for each word of each document a topic is drawn from the mixture and a term is subsequently drawn from the multinomial distribution corresponding to that topicthe topics however evolve over time for instance the two most likely terms of a topic at time matht could be network and zipf in descending order while the most likely ones at time matht could be zipf and percolation in descending ordermodeldefine  as the perdocument topic distribution at time t  as the word distribution of topic k at time t  as the topic distribution for document d in time t  as the topic for the nth word in document d in time t and  as the specific wordin this model the multinomial distributions  respectivelyeven though multinomial distributions are usually written in terms of the mean parameters representing them in terms of the natural parameters is better in the context of dynamic topic modelsthe former representation has some disadvantages due to the fact that the parameters are constrained to be nonnegative and sum to one when defining the evolution of these distributions one would need to assure that such constraints were satisfied since both distributions are in the exponential family one solution to this problem is to represent them in terms of the natural parameters that can assume any real value and can be individually changedusing the natural parameterization the dynamics of the topic model are given byandthe generative process at time slice t is therefore draw topics  draw mixture model  for each document draw  for each word draw topic  draw word where  is a mapping from the natural parameterization x to the mean parameterization namelyinferencein the dynamic topic model only  is observable learning the other parameters constitutes an inference problem blei and lafferty argue that applying gibbs sampling to do inference in this model is more difficult than in static models due to the nonconjugacy of the gaussian and multinomial distributions they propose the use of variational methods in particular the variational kalman filtering and the variational wavelet regressionapplicationin the original paper a dynamic topic model is applied to the corpus of science articles published between  and  aiming to show that this method can be used to analyze the trends of word usage inside topicsref namedtm  the authors also show that the model trained with past documents is able to fit documents of an incoming year better than ldaa continuous dynamic topic model was developed by wang et al and applied to predict the timestamp of documents references"

In [36]:
test_text = "forthe arabic term meaning valleywadiinfobox software name                    web application distribution infrastructure logo                     screenshot              caption                 collapsible             developer               apache software foundation status                  active latest release version  wadi snapshot latest release date     release date latest preview version  latest preview date     operating system        crossplatform size                    programming language    java programming languagejava genre                   license                 apache license  website "

In [41]:
test_matrix_spare = tfidf_vector.transform(sample_text['text'])
test_df_tfd = pd.DataFrame(test_matrix_spare.toarray(),
                                columns=tfidf_vector.get_feature_names())
test_svd_matrix = SVD.transform(test_df_tfd)

In [42]:
lr.predict(test_svd_matrix)

array([1], dtype=uint8)

After testing, the model I have trained predicts solely category 1, Machine Learning. Since this happened multiple times on business software text, the model clearly has data leakage. This is due to the fact that the wikitext I have pulled includes at the least the category the text has been pulled from. The model is likely learning the presence of specific category names and simply predicting on these clear indicators of which top level category the subcategories are part of. 